### PR TITLE
Add benchmarking build step

### PR DESCRIPTION
**Pull request by DRMacIver**
A thing that has annoyed me for a while is that Hypothesis has no benchmarks checked as part of CI.

Unfortunately, timing on Travis is so variable that doing real benchmarking is almost impossible.

Fortunately, as I realised this morning, Hypothesis has an available metric that is completely deterministic (modulo initial random seed) and is an *extremely* good proxy for data generation runtime - the total amount of data actually drawn.

So this PR adds decent statistical benchmarking (albeit for a quite limited set of cases for now) to Hypothesis based on this total amount of data drawn. We run the benchmarks on each build, compare to the known results, and if things have *changed* (not just degraded) the build will fail.

Sadly this will probably slow down our build times a fair bit even though we're only running it in one job. It's not the fastest of tasks, though I won't know how slow it is on Travis until after submitting this.

### CODE REVIEW

**Comment by DRMacIver** [comment]
Looks like for now at least the runtime is actually pretty reasonable.  https://travis-ci.org/HypothesisWorks/hypothesis-python

**Comment by Zac-HD** [comment]
> So this PR adds decent statistical benchmarking (albeit for a quite limited set of cases for now) to Hypothesis based on this total amount of data drawn. We run the benchmarks on each build, compare to the known results, and if things have changed (not just degraded) the build will fail.

Can you explain a little more about how to interpret the statistical results, and why any change is a failure?  In particular, when would you expect or want the results to change and how?

**Comment by DRMacIver** [comment]
> Can you explain a little more about how to interpret the statistical results

Sure (I'll explain in comments here and then update the PR to include this information once it's clear to you).

What we are doing is for each benchmark we are running a T-test for the difference in means of two populations with possibly different variance. The populations we are comparing are a set of runs of Hypothesis under its default settings, counting the total amount of data drawn across all examples.

Each individual benchmark gives us a p-value, which is the probability (under a certain model. I'm not actually sure what scipy uses here but decided to trust their judgement. This may not be valid) that the samples were at least this far apart if the means weren't actually different.

We then run the Benjamini-Hochberg procedure, which uses these p-values to give us a set of benchmarks it believes are anomalous. This procedure controls the *false discovery rate*, which is the expected fraction of the highlighted answers which are false anomalies. We're setting that rate to 1 in 1000, which is fairly conservative (it's a much higher standard than you would typically use in science).

So what comes out at the end of this is a set of benchmarks where we're pretty sure that the mean size generated has changed in some way.

> why any change is a failure?

Change is a failure because if it's an improvement then you should rebuild the improved models using the relevant make task, and if it's a degradation then you should either still update the models or fix it because you've made performance worse.

>  In particular, when would you expect or want the results to change and how?

Most of the time the numbers should go down rather than up. I would expect this to change in response mostly to changes in src/hypothesis/internal/conjecture/engine.py or in changes to the strategy definitions.

File: benchmark-data/intlists-valid=always-interesting=always [hunk 1]
```diff
@@ -0,0 +1,61 @@
+2: 171
```
Comment (froztbyte) [inline benchmark-data/intlists-valid=always-interesting=always]: for posterity's sake:

```
18:50:25 < DRMacIver> froztbyte: Oh, note that that large diff size is mostly autogenerated stuff
18:50:43 < froztbyte> yeah I was about to try comment on some of that
18:51:11 < froztbyte> encoding the properties in only the filename serves the purpose, but leaves something to be desired
18:51:35 < froztbyte> that combined with the provenance of the file content itself leaves me thinking that something else should live either inside or alongside those files
18:51:58 < DRMacIver> I could add a README.rst to that directory?
18:52:03 < froztbyte> that might work, yeah
18:52:07 < DRMacIver> Hmm
18:52:13 < froztbyte> "how these got here, and what the structure is"
18:52:15 < DRMacIver> Actually why don't I add a comment header to the top of the file
18:52:20 < froztbyte> that works too
18:52:28 < froztbyte> and if it's auto-generated, the auto-generator can put it there
18:52:33 < froztbyte> want that as PR comment?
18:52:49 < DRMacIver> froztbyte: Sure
```

File: requirements/benchmark.txt [hunk 2]
```diff
@@ -0,0 +1,9 @@
+appdirs==1.4.3
+attrs==16.3.0
+click==6.7
+hypothesis==3.7.0
```
Comment (froztbyte) [inline requirements/benchmark.txt]: Given that release tooling will auto-increment things, we should have something else in the housekeeping to also take care of this in benchmark suite.

Shall I open a ticket?

File: .pyup.yml [hunk 3]
```diff
@@ -5,3 +5,6 @@ requirements:
   - requirements/test.txt
       updates: all
       pin: True
+  - requirements/benchmark.txt
```
Comment (froztbyte) [inline .pyup.yml]: Ah, ignore previous review comment, pyup will allow us to chain this.

File: scripts/benchmarks.py [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+from scipy.stats import ttest_ind
+import click
+import os
+import random
+import hashlib
+import numpy as np
+import sys
+from collections import Counter
+
+ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
```
Comment (froztbyte) [inline scripts/benchmarks.py]: Same comment here about script path being changed as in #537, although I'm not sure if we want to add a hard dependency on a `git` binary in here. Thoughts?

Comment (froztbyte) [inline scripts/benchmarks.py]: @DRMacIver this is the only review comment worth answering, others are clear. Happy to rely on @Zac-HD's +1 for merge.

Comment (DRMacIver) [inline scripts/benchmarks.py]: I'm broadly OK with making assumptions about the project layout staying the same. Especially in this case I don't think anything particularly bad happens if it changes and this isn't updated.

**Comment by Zac-HD** [comment]
Alright, I like the headers on data files and don't see anything else unaddressed.

**Pull request merged**